cpp/src/fil/cudf_to_matrix.cu

```diff
@@ -0,0 +1,151 @@
+#include "internal.cuh"
+#include <cstdint>
+#include <cudf/column/column_view.hpp>
+#include <cudf/types.hpp>
+#include <raft/handle.hpp>
+#include <thrust/host_vector.h>
+
+template <typename SrcT, typename ScaleT>
+__device__ float convert_maybe_check(const void* in,
+                                     std::size_t idx,
+                                     ScaleT scale,
+                                     int* lossy = nullptr)
+{
+  SrcT orig = reinterpret_cast<const SrcT*>(in)[idx];
+  float res = static_cast<double>(orig) * scale;
+  if (lossy != nullptr) {
+    SrcT back_cast = static_cast<double>(res) / scale;
+    if (back_cast != orig) atomicExch(lossy, 1);
```

Comment:
A volatile write should be enough, using e.g. `cub::ThreadStore<STORE_VOLATILE>()` (see https://nvlabs.github.io/cub/group___util_io.html#ga7e46f6e1c83bab6124cf8d4b66377812).